dcpy/builds/plan.py

```diff
@@ -26,18 +26,25 @@ class RecipeInputsVersionStrategy(str, Enum):
     copy_latest_release = "copy_latest_release"
 
 
-class DataPreprocessor(BaseModel, use_enum_values=True, extra="forbid"):
+class DataPreprocessor(BaseModel, extra="forbid"):
     module: str
     function: str
 
 
+class InputDatasetDestination(str, Enum):
+    postgres = "postgres"
+    df = "df"
+    file = "file"
+
+
 class InputDataset(BaseModel, use_enum_values=True, extra="forbid"):
     name: str
     version: str | None = None
     file_type: recipes.DatasetType | None = None
     version_env_var: str | None = None
     import_as: str | None = None
     preprocessor: DataPreprocessor | None = None
+    destination: InputDatasetDestination | None = None
```

Comment:
All the nullability here makes me think we probably need to either 
1. implement a custom deserializer, or
3. have separate classes to differentiate between what we've serialized from yaml, and what we've actually resolved. So maybe this class should be called a InputDatasetRaw where everything is nullable, then we have another class which is stricter. That would allow us to remove the assertions you've added, I think?

Playing around with how this could work, and it could look something like this: 
```
class Strict(BaseModel):
    name: str
    version: str

class Lax(Strict):
    version: str | None = None # Only override this field, to loosen the constraint. 
    
l = Lax(name="hi", version="there") # used for deserializing
s = Strict() # used for resolved. Throws exception bc missing name and version
```
The inheritance direction feels backwards somehow. My gut says you'd want to subclass to make something _more_ specific (ie remove nullability) rather than open it up. But such is life in OOP-land.

Reply:
My gut is that if we went with separate classes, might be best to avoid the inheritance route, and just have them be fully distinct. So you'd only have a strict when expecting a strict, only have a lax when expecting a lax. Issue then is we need to go up the chain so to speak, have "lax" recipes and "computed"/strict recipes as separate classes or something like that. Not necessarily a bad thing, but a bit boilerplate-y. But might be worth it for these sorts of situations.

Isn't that a bit orthogonal from the custom deserializer though? Not entirely sure how you mean we would use it to get around the `| None`s

Reply:
Re: the custom deserializer, I was thinking that a large part of why we allow nullability for these class attributes is that we omit values in the serialized data (ie the yaml files) then immediately compute them after we deserialize. e.g. Setting the filetype of a dataset object. So if we were able to hook into the deserializer, we could avoid those types of situations potentially. It's probably my less preferred option though.

Reply:
Ah I see what you mean. Yeah I'd prefer to not handle that sort of thing during deserialization, that seems like it'd get messy in a hurry